src/remote/fs/fs.go

```diff
@@ -0,0 +1,210 @@
+// Package fs provides an io/fs.FS implementation over the remote API
+package fs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	iofs "io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/client"
+	"github.com/bazelbuild/remote-apis-sdks/go/pkg/digest"
+	pb "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"google.golang.org/protobuf/proto"
+)
+
+type Client interface {
+	ReadBlob(ctx context.Context, d digest.Digest) ([]byte, *client.MovedBytesMetadata, error)
+}
+
+// fs is an io/fs.FS implemented on top of a REAPI directory. This will download files on demand, as they are needed.
+type fs struct {
+	c    Client
+	root *pb.Directory
+}
+
+// New creates a new filesystem on top of the given proto, using client to download files on demand.
+func New(c Client, root *pb.Directory) iofs.FS {
+	return &fs{
+		c:    c,
+		root: root,
+	}
+}
+
+// Open opens the file with the given name
+func (fs *fs) Open(name string) (iofs.File, error) {
+	return fs.open(".", filepath.Clean(name), fs.root)
+}
+
+func (fs *fs) open(path, name string, wd *pb.Directory) (iofs.File, error) {
+	name, rest, hasToBeDir := strings.Cut(name, string(filepath.Separator))
+	// Must be a dodgy symlink that goes past our root.
+	if name == ".." || name == "." {
+		return nil, os.ErrNotExist
+	}
+
+	for _, d := range wd.Directories {
+		if d.Name == name {
+			dg, err := digest.NewFromProto(d.Digest)
+			if err != nil {
+				return nil, err
+			}
+			bs, _, err := fs.c.ReadBlob(context.Background(), dg)
+			if err != nil {
+				return nil, err
+			}
+			dirPb := &pb.Directory{}
+			if err := proto.Unmarshal(bs, dirPb); err != nil {
+				return nil, err
+			}
+			if rest == "" {
+				return &dir{
+					info: info{
+						mode:    iofs.FileMode(dirPb.NodeProperties.UnixMode.Value),
+						modTime: dirPb.NodeProperties.GetMtime().AsTime(),
+						name:    name,
+						isDir:   true,
+					},
+					pb: dirPb,
+				}, nil
+			}
+			return fs.open(filepath.Join(path, name), rest, dirPb)
+		}
+	}
+
+	// If the path contains a /, we only resolve against dirs.
+	if hasToBeDir {
+		return nil, iofs.ErrNotExist
+	}
+
+	for _, f := range wd.Files {
+		if f.Name == name {
+			d, err := digest.NewFromProto(f.Digest)
+			if err != nil {
+				return nil, err
+			}
+			bs, _, err := fs.c.ReadBlob(context.Background(), d)
+			if err != nil {
+				return nil, err
+			}
+			return &file{
+				bs: bs,
+				info: info{
+					isDir:   false,
+					size:    int64(len(bs)),
+					mode:    iofs.FileMode(f.NodeProperties.UnixMode.Value),
+					modTime: f.NodeProperties.GetMtime().AsTime(),
+					name:    f.Name,
+				},
+			}, nil
+		}
+	}
+
+	for _, l := range wd.Symlinks {
+		if l.Name == name {
+			if filepath.IsAbs(l.Target) {
+				// The doc comments suggest that sometimes this is supported. I'm not sure how this would be useful in
```

Comment:
I'm not sure that we ever really allow this. IIRC there are some flags about whether the server supports/allows it but it seems like a Bad Idea to me

Reply:
Cool cool, I will update the comment.